### PR TITLE
Do not run clone step if no pipeline step will run

### PR DESCRIPTION
**Pull request by qwerty287**
Skip the clone step and ignore hook/pipeline if no pipeline step except clone would run. The status reported back to the forge is `success`.

Closes https://github.com/woodpecker-ci/woodpecker/issues/778

### CODE REVIEW

**Comment by anbraten** [comment]
> With one pipeline, an issue occurs: the time says not started yet because the duration is 0 which results in displaying this message. I don't know a simple way to fix it. Also, the large white "box" next to the pipeline list looks a bit odd.

I improved the "white box" a bit in #846. Maybe that adjustment would be helpful for you in this PR as well.

What do you think about hiding empty pipelines and maybe the complete "build" in case they are empty as we are doing a similar thing for pipelines steps which are not executed => they are not shown to the user as well?

**Comment by qwerty287** [comment]
Yeah I thought about this too but wasn't able to implement it after some experiments (I don't know any web frontend dev, which I'd need to add such a weird `v-if` statement to the vue files, sorry). Hiding the complete build would be possible, then an easy solution would also to just return if coming from the hook (i.e. don't initiate a build). This wouldn't report a status back to the forge, which can be problematic if using pull requests that require the checks. Hinding it only from the list is a good idea, but would require more effort which I'm not able to do because I don't understand woodpecker's structure enough (and don't vue in general).

**Comment by 6543** [comment]
I would: if there are no steps, dont add clone but report success back

and only if global pipeline filter #when (#283 ) do filter out make no response

File: agent/runner.go [hunk 1]
```diff
@@ -75,6 +75,16 @@ func (r *Runner) Run(ctx context.Context) error {
 		return nil
 	}
 
+	if len(work.Config.Stages) < 1 {
```
Comment (6543) [inline agent/runner.go:78]: I dont think this should affect the agent at all - it's a pipeline compiler thing

Comment (anbraten) [reply agent/runner.go:78]: You are right it should work without.

**Comment by 6543** [comment]
> After the commits I pushed now it just skips and ignores the hook. Should it send a "success" status back and should this always happen for empty builds?

If no steps but pipeline is not filtered golbaly -> success
if pipline is filtered globaly -> none

File: server/api/hook.go [hunk 2]
```diff
@@ -200,10 +200,18 @@ func PostHook(c *gin.Context) {
 		return
 	}
 
-	if zeroSteps(build, remoteYamlConfigs) {
+	if z, steps := zeroSteps(build, remoteYamlConfigs); z {
```
Comment (6543) [inline server/api/hook.go:203]:
```suggestion
	// TODO: move global pipeline filters into own check functions ...
	if z, steps := zeroSteps(build, remoteYamlConfigs); z {
```

File: server/api/hook.go [hunk 3]
```diff
@@ -200,10 +200,19 @@ func PostHook(c *gin.Context) {
...
 		msg := "ignoring hook: step conditions yield zero runnable steps"
 		log.Debug().Str("repo", repo.FullName).Msg(msg)
 		c.String(http.StatusOK, msg)
+		build.Status = model.StatusSuccess
+		for _, step := range steps {
+			step.Proc.State = model.StatusSuccess
+			build.Procs = append(build.Procs, step.Proc)
+		}
+		if err := updateBuildStatus(c, build, repo, repoUser); err != nil {
```
Comment (6543) [inline server/api/hook.go:213]:
```suggestion
		// TODO: this wont create any builds the status can link to ...
		if err := updateBuildStatus(c, build, repo, repoUser); err != nil {
```

**Review by 6543 — APPROVED** [review]
some todos ... but that should be covered in another pull (refactoring etc ...)

**Pull request merged**